src/Utilities/Functional.hpp

```diff
@@ -0,0 +1,155 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <cmath>
+#include <cstddef>
+#include <tuple>
+
+/*!
+ * \ingroup UtilitiesGroup
+ * \brief Higher order function objects similar to `std::plus`, etc.
+ */
+namespace funcl {
+/// \cond
+template <size_t Arity>
+struct Functional {
+  static constexpr size_t arity = Arity;
+
+ protected:
+  template <class C, size_t Offset, class... Ts, size_t... Is>
+  static constexpr decltype(auto) helper(
+      const std::tuple<Ts...>& t,
+      std::index_sequence<Is...> /*meta*/) noexcept {
+    return C{}(std::get<Offset + Is>(t)...);
+  }
+};
+
+struct Identity;
+/// \endcond
+
+#define MAKE_BINARY_OPERATOR(NAME, OPERATOR)                            \
+  template <class C0 = Identity, class C1 = C0>                         \
+  struct NAME : Functional<C0::arity + C1::arity> {                     \
+    using base = Functional<C0::arity + C1::arity>;                     \
+    template <class... Ts>                                              \
+    constexpr auto operator()(const Ts&... ts) noexcept {               \
+      return base::template helper<C0, 0>(                              \
+          std::tuple<const Ts&...>(ts...),                              \
+          std::make_index_sequence<C0::arity>{})                        \
+          OPERATOR base::template helper<C1, C0::arity>(                \
+              std::tuple<const Ts&...>(ts...),                          \
+              std::make_index_sequence<C1::arity>{});                   \
+    }                                                                   \
+  };                                                                    \
+  /** \cond */                                                          \
+  template <class C1>                                                   \
+  struct NAME<Identity, C1> : Functional<1 + C1::arity> {               \
+    template <class T0, class... Ts>                                    \
+    constexpr auto operator()(const T0& t0, const Ts&... ts) noexcept { \
+      return t0 OPERATOR C1{}(ts...);                                   \
+    }                                                                   \
+  };                                                                    \
+  template <>                                                           \
+  struct NAME<Identity, Identity> : Functional<2> {                     \
+    template <class T0, class T1>                                       \
+    constexpr auto operator()(const T0& t0, const T1& t1) noexcept {    \
+      return t0 OPERATOR t1;                                            \
+    }                                                                   \
+  };                                                                    \
+  /** \endcond*/
+
+#define MAKE_BINARY_FUNCTIONAL(NAME, OPERATOR)                           \
```

Comment:
This macro is never used.

Reply:
Do you want it removed or add some operator that uses it?

Reply:
I just removed it :)

Reply:
Remove the two other mentions of it as well.

Reply:
done